File: C++/Sequences/(20)ValidParentheses.cpp
// Stack
// Time: O(n)
// Space: O(n)
class Solution {
public:
    bool isValid(string s) {
        stack<char> values;
        
        for (int i = 0; i < s.length(); i++) {
            if (s[i] == '(' || s[i] == '{' || s[i] == '[') {
                values.push(s[i]);
            }
            else {
                if (values.size() == 0)
                    return false;
                if (s[i] == ')' && values.top() != '(')
                    return false;
                if (s[i] == '}' && values.top() != '{')
                    return false;
                if (s[i] == ']' && values.top() != '[')
                    return false;
                values.pop();
            }
        }
        
        return values.empty();
    }
};

File: Python/Sequences/(20)ValidParentheses.py
# Stack
# Time: O(n)
# Space: O(n)
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        
        for i in range(len(s)):
            if s[i] == '(' or s[i] == '[' or s[i] == '{':
                stack.append(s[i])
            else:
                if len(stack) == 0:
                    return False
                if s[i] == ')' and stack[len(stack) - 1] != '(':
                    return False
                if s[i] == ']' and stack[len(stack) - 1] != '[':
                    return False
                if s[i] == '}' and stack[len(stack) - 1] != '{':
                    return False
                stack.pop()
        
        return not stack